### PR TITLE
Mistral vulkan lib for win/linux f16/f32 with sw=4k cs=128

### PR DESCRIPTION
**Pull request by Sing-Li**
get Mistral to work on common configs

### CODE REVIEW

**Comment by bymihaj** [comment]
Hello.
I have test you Mistral-7B-Instruct-v0.1-q4f32_1-vulkan.dll on Windows and look like input message with size more than 230 symbol throws error in my system.

python -m mlc_chat.rest --model Mistral-7B-Instruct-v0.1-q4f32_1
INFO:     Started server process [13124]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
INFO:     Uvicorn running on http://127.0.0.1:8000 (Press CTRL+C to quit)
[11:37:55] D:\a\package\package\tvm\src\runtime\memory\memory_manager.cc:108: InternalError: Check failed: (offset + needed_size <= this->buffer.size) is false: storage allocation failure, attempted to allocate 2113536 at offset 0 in region that is 2097152bytes
Stack trace not available when DMLC_LOG_STACK_TRACE is disabled at compile time.

ERROR:    Exception in ASGI application
Traceback (most recent call last):
  File "C:\Users\mihaj\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\site-packages\uvicorn\protocols\http\h11_impl.py", line 428, in run_asgi
    result = await app(  # type: ignore[func-returns-value]
  File "C:\Users\mihaj\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\site-packages\uvicorn\middleware\proxy_headers.py", line 78, in __call__
    return await self.app(scope, receive, send)
  File "C:\Users\mihaj\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\site-packages\fastapi\applications.py", line 281, in __call__
    await super().__call__(scope, receive, send)
  File "C:\Users\mihaj\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\site-packages\starlette\applications.py", line 122, in __call__
    await self.middleware_stack(scope, receive, send)
  File "C:\Users\mihaj\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\site-packages\starlette\middleware\errors.py", line 184, in __call__
    raise exc
  File "C:\Users\mihaj\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\site-packages\starlette\middleware\errors.py", line 162, in __call__
    await self.app(scope, receive, _send)
  File "C:\Users\mihaj\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\site-packages\starlette\middleware\cors.py", line 83, in __call__
    await self.app(scope, receive, send)
  File "C:\Users\mihaj\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\site-packages\starlette\middleware\exceptions.py", line 79, in __call__
    raise exc
  File "C:\Users\mihaj\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\site-packages\starlette\middleware\exceptions.py", line 68, in __call__
    await self.app(scope, receive, sender)
  File "C:\Users\mihaj\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\site-packages\fastapi\middleware\asyncexitstack.py", line 20, in __call__
    raise e
  File "C:\Users\mihaj\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\site-packages\fastapi\middleware\asyncexitstack.py", line 17, in __call__
    await self.app(scope, receive, send)
  File "C:\Users\mihaj\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\site-packages\starlette\routing.py", line 718, in __call__
    await route.handle(scope, receive, send)
  File "C:\Users\mihaj\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\site-packages\starlette\routing.py", line 276, in handle
    await self.app(scope, receive, send)
  File "C:\Users\mihaj\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\site-packages\starlette\routing.py", line 66, in app
    response = await func(request)
  File "C:\Users\mihaj\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\site-packages\fastapi\routing.py", line 273, in app
    raw_response = await run_endpoint_function(
  File "C:\Users\mihaj\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\site-packages\fastapi\routing.py", line 190, in run_endpoint_function
    return await dependant.call(**values)
  File "C:\Users\mihaj\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\site-packages\mlc_chat\rest.py", line 241, in request_chat_completion
    msg = session["chat_mod"].generate(prompt=request.messages, generation_config=generation_config)
  File "C:\Users\mihaj\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\site-packages\mlc_chat\chat_module.py", line 856, in generate
    self._prefill(prompt, generation_config=generation_config)
  File "C:\Users\mihaj\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\site-packages\mlc_chat\chat_module.py", line 1073, in _prefill
    self._prefill_func(
  File "C:\Users\mihaj\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\site-packages\tvm\_ffi\_ctypes\packed_func.py", line 239, in __call__
    raise_last_ffi_error()
  File "C:\Users\mihaj\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\site-packages\tvm\_ffi\base.py", line 481, in raise_last_ffi_error
    raise py_err
tvm._ffi.base.TVMError: Traceback (most recent call last):
  File "D:\a\package\package\tvm\src\runtime\memory\memory_manager.cc", line 108
InternalError: Check failed: (offset + needed_size <= this->buffer.size) is false: storage allocation failure, attempted to allocate 2113536 at offset 0 in region that is 2097152bytes

**Comment by Sing-Li** [comment]
Looks like you're running out of memory on your GPU.  Which GPU are you running on?  @bymihaj

**Comment by bymihaj** [comment]
RX 6800 with 16G VRAM. And only 6 Gb is used according Task Manager. Additionally, model is able to produce long output - up 2.000 tokens. Issue is about only long input.

**Comment by Sing-Li** [comment]
@davidpissarra  do you know of any recent bug in mistral  sw=4k cs=128 handling that may cause problem with input token > 230 tokens ?     I do think a rebuild will do anything for @bymihaj      Thanks.

**Comment by davidpissarra** [comment]
Hi @bymihaj, can you confirm the `mlc-chat-config.json` file has the correct parameters (so it matches the library sw=4k cs=128):

```
"sliding_window": 4096,
"prefill_chunk_size": 128

**Comment by bymihaj** [comment]
@davidpissarra
I did not modify config. So it is https://huggingface.co/mlc-ai/mlc-chat-Mistral-7B-Instruct-v0.1-q4f32_1/commit/73cbe86c36adc3c699d77681746794810795a579#d2h-124584 ( but it was previous version)

So "sliding_window": 4096, but prefill_chunk_size:  has not been defined. That was fixed  in https://huggingface.co/mlc-ai/mlc-chat-Mistral-7B-Instruct-v0.1-q4f32_1/commit/4e48c68a139c4693e09926abc87e71c3e43bb13a
So.. maybe still wrong value. 

When I updated (pull) config from HF, I get new error on start-up 
File "D:\a\package\package\mlc-llm\cpp\llm_chat.cc", line 413
TVMError: Check failed: (config.count("sliding_window_chunk_size")) is false: Need to specify chunk size if using sliding window attention.

**Comment by Sing-Li** [comment]
@bymihaj   the latest code checks "prefill_chunk_size"    -- please verify that you have the latest code.

Otherwise, try adding:

```
"sliding_window_chunk_size" : 4096
```

That should at least get past your error.

**Comment by davidpissarra** [comment]
@bymihaj let me know if it works with the latest code. As mentioned by @Sing-Li, as of https://github.com/mlc-ai/mlc-llm/pull/1280 `sliding_window_chunk_size` was renamed to `prefill_chunk_size` to generalize chunking to the remaining non-SWA models.

**Comment by bymihaj** [comment]
Solved. I have updated https://mlc.ai/wheels mlc-chat-nightly mlc-ai-nightly and set "prefill_chunk_size": 128. It is working on my PC with long input. I think config into HF might be updated as prefill_chunk_size is 4096 there.

**Comment by tqchen** [comment]
@davidpissarra please check again and see if we can merge this

File: Mistral-7B-Instruct-v0.1-q4f32_1-sw4096_cs128-vulkan.so
Comment (davidpissarra) [inline Mistral-7B-Instruct-v0.1-q4f32_1-sw4096_cs128-vulkan.so:1]: I wasn't able to run this one @Sing-Li

File: Mistral-7B-Instruct-v0.1-q4f32_1-sw4096_cs128-cuda.so
Comment (davidpissarra) [inline Mistral-7B-Instruct-v0.1-q4f32_1-sw4096_cs128-cuda.so:1]: This lib seems to have 2k of sliding window

**Review by davidpissarra — COMMENTED** [review]
Hi @Sing-Li can you rebuild the libraries with the correct parameters and the latest code. `prefill_chunk_size` is missing in the metadata, so rebuilding with the latest code should be enough. Thanks a lot!

File: Mistral-7B-Instruct-v0.1-q4f32_1-sw4096_cs128-vulkan.so
Comment (Sing-Li) [reply Mistral-7B-Instruct-v0.1-q4f32_1-sw4096_cs128-vulkan.so:1]: @davidpissarra    It worked fine for me - but I have to rename the `.so`, see this run (latest nightly and latest huggingface weights as of today):

```
Use MLC config: "/home/autoqa/dev/mlc-chat/dist/prebuilt/mlc-chat-Mistral-7B-Instruct-v0.1-q4f32_1/mlc-chat-config.json"
Use model weights: "/home/autoqa/dev/mlc-chat/dist/prebuilt/mlc-chat-Mistral-7B-Instruct-v0.1-q4f32_1/ndarray-cache.json"
Use model library: "/home/autoqa/dev/mlc-chat/dist/prebuilt/lib/Mistral-7B-Instruct-v0.1-q4f32_1-vulkan.so"
You can use the following special commands:
  /help               print the special commands
  /exit               quit the cli
  /stats              print out the latest stats (token/sec)
  /reset              restart a fresh chat
  /reload [model]  reload model `model` from disk, or reload the current model if `model` is not specified

Loading model...
Loading finished
Running system prompts...
System prompts finished
[INST]: what are the states in the usa?
[/INST]: There are 50 states in the United States. They are:

1. Alabama
2. Alaska
3. Arizona
4. Arkansas
5. California
6. Colorado
7. Connecticut
8. Delaware
```

File: Mistral-7B-Instruct-v0.1-q4f32_1-sw4096_cs128-cuda.so
Comment (Sing-Li) [reply Mistral-7B-Instruct-v0.1-q4f32_1-sw4096_cs128-cuda.so:1]: rebuild with latest nightly and ensured `--sliding-window 4096 --prefill-chunk-size 128` during build ; tested with latest huggingface weights

Please test.

**Comment by tqchen** [comment]
closing for now as we can use JIT for vulkan cases

**Pull request closed without merging**